enhancements/update/upgrades-blocking-on-ack.md

```diff
@@ -0,0 +1,181 @@
+---
+title: upgrade-acknowledgment-gate
+authors:
+  - "@jottofar"
+reviewers:
+  - "@LalatenduMohanty"
+  - "@bparees"
+  - "@sdodson"
+  - "@csrwng"
+  - "@relyt0925"
+  - "@jeremyeder"
+approvers:
+  - TBD 
+creation-date: 2021-07-08
+last-updated: 2021-07-14
+status: implementable
+---
+
+# Upgrade Acknowledgment Gate
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+CVO currently blocks minor level upgrades when overrides are set, a resource deletion is pending, or an operator has set its upgradeable condition to false.
+This enhancement adds a generic upgrade blocking mechanism for blocking minor level upgrades.
+This generic upgrade blocker can consist of one or more items each of which will block a minor level upgrade.
+Each generic upgrade blocker item will be associated with some well-defined task to be completed by an administrator.
+Once the task has been completed the administrator will acknowledge completion and its associated item will no longer block minor level upgrades.
+
+## Motivation
+
+OCP 4.9 uses kubernetes v1.22 which is removing a large number of [APIs][k8s-1.22-removed-apis].
+We suspect that many customer clusters are still using these APIs through helm charts, operators, generic workloads, administrative tools, etc.
+We believe there is no effective way for a Cluster Operator to detect the use of these APIs by external applications.
+Relying solely on automated API detection to block upgrades could result in blocking for APIs that are not actively used any longer or not blocking because an application was not active during the check.
+Having a human in the loop to drive the analysis and make the final decision is the surest way to prevent customers from upgrading from 4.8.z to 4.9.z and encountering missing API related breakage.
+
+### Goals
+
+* Acknowledgment requires the explicit update of a cluster resource by an administrator.
+* For the initial release in 4.8.z, administrator acknowledgment of API removal will be required before a minor level upgrade to 4.9 is allowed.
+* Upgrade blocking mechanism must be generic such that it can be repurposed in the future.
+* It must be clear what an administrator is acknowledging.
+* Administrators can bypass the upgrade blocking mechanism by using the existing CVO "force" implementation.
+* The UI will clearly indicate when this upgrade blocking mechanism is blocking upgrades, why, and what needs to be done to clear the block.
+* This upgrade blocking mechanism must be compatible with SD, ACM, IKS/ROKS, Hive, and Hypershift.
```

Comment:
Hive doesn't really deal with upgrades today but it's something we've always talked about.

@jnpacker what do you think, does this fit best in ACM or Hive?

Will there be a programatic way for a tool initiating the upgrade to query the ack's required for an upcoming release image the cluster may upgrade to? Something in the release image ideally.

Reply:
> Will there be a programatic way for a tool initiating the upgrade to query the ack's required for an upcoming release image the cluster may upgrade to? Something in the release image ideally.

there isn't one currently defined, no.  When you say something in the release image, what do you have in mind?

Reply:
some of kind of manifest that something like ACM or Hive could parse to put infront of the user, here are the things you need to opt-in for before this cluster will upgrade. 

Actually that leads to some questions, how does the CVO know what keys it needs? Do we backport what is needed for 4.9 into 4.8.z? Or is the CVO actively examining possible upgrade endpoints and again, parsing out the keys, so it knows it's not upgradable. (apologies if I missed this)

Reply:
The gating logic will go into 4.8.z, not 4.9 (for this particular gate anyway).  Right now the plan is to just code the logic in the CVO itself.  We had an earlier iteration where the gates were defined via resources, but it started to get complicated.

Reply:
How could ACM/Hive know what flags need to be set?

Reply:
honestly my thought was the ACM would be specifically aware of the acks a particular OCP update requires and include code accordingly.

but we are discussing moving the list of required gates into a configmap again.

Reply:
Based on team discussion added additional configmap (for a total of two new configmaps) to contain keys.

Reply:
This looks good, thanks!